element-connector/src/main/java/org/eclipse/californium/elements/tcp/TcpServerConnector.java

```diff
@@ -82,17 +85,17 @@ public TcpServerConnector(InetSocketAddress localAddress, int numberOfThreads, i
 				.option(ChannelOption.AUTO_READ, true).childHandler(new ChannelRegistry());
 
 		// Start the server.
-		bootstrap.bind(localAddress).syncUninterruptibly().addListener(new ChannelFutureListener() {
-
-			@Override
-			public void operationComplete(ChannelFuture future) throws Exception {
-				InetSocketAddress listenAddress = (InetSocketAddress) future.channel().localAddress();
-				listenUri = getListenUri(listenAddress);
-			}
-		});
+		ChannelFuture channelFuture = bootstrap.bind(localAddress).syncUninterruptibly();
+
+		if (channelFuture.isSuccess() && 0 == localAddress.getPort()) {
+			InetSocketAddress listenAddress = (InetSocketAddress) channelFuture.channel().localAddress();
+			localAddress = new InetSocketAddress(localAddress.getAddress(), listenAddress.getPort());
```

Comment:
Using that results in:

     coap+tcp://0:0:0:0:0:0:0:0:5683

and using that as (leshan example):

        this.registrationEndpointUri = registrationEndpointUri;
        this.registrationEndpointAddress = new InetSocketAddress(registrationEndpointUri.getHost(), registrationEndpointUri.getPort());

results in:

    java.lang.IllegalArgumentException: hostname can't be null
	at java.net.InetSocketAddress.checkHost(InetSocketAddress.java:149) ~[?:1.8.0_51]
	at java.net.InetSocketAddress.<init>(InetSocketAddress.java:216) ~[?:1.8.0_51]

So it seems, IPv6 has some open issues. Therefore I just replaced the port.
I had a comment on that, but on the reworks that comment seems to be deleted.
Alternatively, IPv6 usage may be consequently supported. But this seems to be much out of this scope.